raft/src/main/java/org/apache/kafka/raft/internals/UpdateVoterHandler.java

```diff
@@ -0,0 +1,233 @@
+import org.apache.kafka.common.protocol.Errors;
+import org.apache.kafka.common.utils.Time;
+import org.apache.kafka.raft.Endpoints;
+import org.apache.kafka.raft.LeaderAndEpoch;
+import org.apache.kafka.raft.LeaderState;
+import org.apache.kafka.raft.LogOffsetMetadata;
+import org.apache.kafka.raft.RaftUtil;
+import org.apache.kafka.server.common.KRaftVersion;
+
+import java.util.Optional;
+import java.util.OptionalInt;
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * This type implements the protocol for updating a voter from a KRaft partition.
+ *
+ * 1. Check that the leader has fenced the previous leader(s) by checking that the HWM is known,
+ *    otherwise return the REQUEST_TIMED_OUT error.
+ * 2. Check that the cluster supports kraft.version 1, otherwise return the UNSUPPORTED_VERSION error.
+ * 3. Check that there are no uncommitted voter changes, otherwise return the REQUEST_TIMED_OUT error.
+ * 4. Check that the updated voter still supports the currently finalized kraft.version, otherwise
+ *    return the INVALID_REQUEST error.
+ * 5. Check that the updated voter is still listening on the default listener.
+ * 6. Append the updated VotersRecord to the log. The KRaft internal listener will read this
+ *    uncommitted record from the log and update the voter in the set of voters.
+ * 7. Wait for the VotersRecord to commit using the majority of the voters. Return a
+ *    REQUEST_TIMED_OUT error if it doesn't commit in time.
+ * 8. Send the UpdateVoter successful response to the voter.
+ *
+ * KAFKA-16538 is going to add support for handling this RPC when the kraft.version is 0.
+ */
+public final class UpdateVoterHandler {
+    private final OptionalInt localId;
+    private final KRaftControlRecordStateMachine partitionState;
+    private final ListenerName defaultListenerName;
+    private final Time time;
+    private final long requestTimeoutMs;
+
+    public UpdateVoterHandler(
+        OptionalInt localId,
+        KRaftControlRecordStateMachine partitionState,
+        ListenerName defaultListenerName,
+        Time time,
+        long requestTimeoutMs
+    ) {
+        this.localId = localId;
+        this.partitionState = partitionState;
+        this.defaultListenerName = defaultListenerName;
+        this.time = time;
+        this.requestTimeoutMs = requestTimeoutMs;
+    }
+
+    public CompletableFuture<UpdateRaftVoterResponseData> handleUpdateVoterRequest(
+        LeaderState<?> leaderState,
+        ListenerName requestListenerName,
+        ReplicaKey voterKey,
+        Endpoints voterEndpoints,
+        UpdateRaftVoterRequestData.KRaftVersionFeature supportedKraftVersions,
+        long currentTimeMs
+    ) {
+        // Check if there are any pending voter change requests
+        if (leaderState.isOperationPending(currentTimeMs)) {
+            return CompletableFuture.completedFuture(
+                RaftUtil.updateVoterResponse(
+                    Errors.REQUEST_TIMED_OUT,
```

Comment:
another place where it would be nice if there was an error msg (or at least logging)